mlflow/store/tracking/file_store.py

```diff
@@ -1319,3 +1326,363 @@ def _read_helper(root, file_name, attempts_remaining=2):
+        """
+        trace_info, trace_dir = self._get_trace_info_and_dir(request_id)
+        if key not in trace_info.tags:
+            raise MlflowException(
+                f"No tag with name: {key} in trace with request_id {request_id}.",
+                RESOURCE_DOES_NOT_EXIST,
+            )
+        trace_info.tags.pop(key)
+        _validate_tag_name(key)
+        tag_path = os.path.join(trace_dir, FileStore.TRACE_TAGS_FOLDER_NAME, key)
+        if not exists(tag_path):
+            raise MlflowException(
+                f"No tag with name: {key} in trace with request_id {request_id}.",
+                RESOURCE_DOES_NOT_EXIST,
+            )
+        os.remove(tag_path)
+
+    def delete_traces(
+        self,
+        experiment_id: str,
+        max_timestamp_millis: Optional[int] = None,
+        max_traces: Optional[int] = None,
+        request_ids: Optional[List[str]] = None,
+    ) -> int:
+        """
+        Delete traces based on the specified criteria.
+
+        - Either `max_timestamp_millis` or `request_ids` must be specified, but not both.
+        - `max_traces` can't be specified if `request_ids` is specified.
+
+        Args:
+            experiment_id: ID of the associated experiment.
+            max_timestamp_millis: The maximum timestamp in milliseconds since the UNIX epoch for
+                deleting traces. Traces older than or equal to this timestamp will be deleted.
+            max_traces: The maximum number of traces to delete.
+            request_ids: A set of request IDs to delete.
+
+        Returns:
+            The number of traces deleted.
+        """
+        # request_ids can't be an empty list of string
+        if max_timestamp_millis is None and not request_ids:
+            raise MlflowException(
+                "Either `max_timestamp_millis` or `request_ids` must be specified.",
+                INVALID_PARAMETER_VALUE,
+            )
+        if max_timestamp_millis and request_ids:
+            raise MlflowException(
+                "Only one of `max_timestamp_millis` and `request_ids` can be specified.",
+                INVALID_PARAMETER_VALUE,
+            )
+        if request_ids and max_traces is not None:
+            raise MlflowException(
+                "`max_traces` can't be specified if `request_ids` is specified.",
+                INVALID_PARAMETER_VALUE,
+            )
+        if max_traces is not None and max_traces <= 0:
+            raise MlflowException(
+                f"`max_traces` must be a positive integer, received {max_traces}.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        experiment_path = self._get_experiment_path(experiment_id, assert_exists=True)
+        traces_path = os.path.join(experiment_path, FileStore.TRACES_FOLDER_NAME)
+        deleted_traces = 0
+        if max_timestamp_millis:
+            trace_paths = list_all(traces_path, lambda x: os.path.isdir(x), full_path=True)
+            if max_traces is None:
```

Comment:
WDYT about a validation step to alert users that they're requesting to delete more traces than exist? 
For instance, I start logging traces in my experiment. I log 10 of them. Then I start a new experiment and log 5000 traces. I then call the delete_traces API and set 4000 as the max_traces, accidentally putting the first experiment ID in there. All traces for that experiment are lost. 
Do you think we should alert users if they're calling for explicitly deleting more traces than exist for a given experiment_id?

Reply:
Actually I think we don't need to warn them, it's possible that users don't know how many traces they have in total so they just specify a large number to make sure all the traces are deleted. If there's a FR to "recover" traces user has deleted, we could support `trace stage -- DELETED` afterwards, but in general I feel like the `max_traces` name is already informative and users should be aware of what this means.

Reply:
+1 to no warning.